### ISSUE TITLE
Sync: Memory violation on Raspberry Pi 4b

### ISSUE THREAD

**Issue by robinaly** [rclone/rclone#5820]
#### The associated forum post URL from `https://forum.rclone.org`

None

#### What is the problem you are having with rclone?

An apparently valid call leads to memory violations

#### What is your rclone version (output from `rclone version`)

rclone v1.57.0
- os/version: raspbian 10.10
- os/kernel: 5.10.52-v7l+ (armv7l)
- os/type: linux
- os/arch: arm
- go/version: go1.17.2
- go/linking: static
- go/tags: none

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Linux, Raspberry Pi 4b, ARM7l

#### Which cloud storage system are you using? (e.g. Google Drive)

Microsoft Blob Storage

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

rclone-v1.57.0-linux-arm-v7/rclone sync -vv test/ porkvision:/

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
2021/11/16 11:40:40 DEBUG : rclone: Version "v1.57.0" starting with parameters ["rclone-v1.57.0-linux-arm-v7/rclone" "sync" "-vv" "test/" "porkvision:/"]
2021/11/16 11:40:40 DEBUG : Creating backend with remote "test/"
2021/11/16 11:40:40 DEBUG : Using config file from "/home/pi/.config/rclone/rclone.conf"
2021/11/16 11:40:40 DEBUG : fs cache: renaming cache item "test/" to be canonical "/home/pi/test"
2021/11/16 11:40:40 DEBUG : Creating backend with remote "porkvision:/"
2021/11/16 11:40:40 DEBUG : fs cache: renaming cache item "porkvision:/" to be canonical "porkvision:"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x56e308]

goroutine 32 [running]:
github.com/Azure/azure-storage-blob-go/azblob.blockBlobClient.CommitBlockList({{{{0x0, 0x0}, {0x0, 0x0}, 0x0, {0x0, 0x0}, {0x376dd70, 0x7}, {0x0, ...}, ...}, ...}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/zz_generated_block_blob.go:82 +0x2f4
github.com/Azure/azure-storage-blob-go/azblob.BlockBlobURL.CommitBlockList({{{{{...}, {...}}}}, {{{{...}, {...}, 0x0, {...}, {...}, {...}, 0x0, {...}, ...}, ...}}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/url_block_blob.go:120 +0x1c0
github.com/Azure/azure-storage-blob-go/azblob.(*copier).close(0x34b01c0)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/chunkwriting.go:186 +0x118
github.com/Azure/azure-storage-blob-go/azblob.copyFromReader({0x180d580, 0x36f0f00}, {0x17f501c, 0x351efd0}, {0x1807564, 0x34f0280}, {{0x180d760, 0x37fa1f0}, 0x0, 0x400000, ...})
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/chunkwriting.go:64 +0x24c
github.com/Azure/azure-storage-blob-go/azblob.UploadStreamToBlockBlob({0x180d580, 0x36f0f00}, {0x17f501c, 0x351efd0}, {{{{{...}, {...}}}}, {{{{...}, {...}, 0x0, {...}, ...}, ...}}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/highlevel.go:553 +0xf0
github.com/rclone/rclone/backend/azureblob.(*Object).Update.func1()
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1680 +0x200
github.com/rclone/rclone/fs.pacerInvoker(0x1, 0x1, 0x397e3f0)
	github.com/rclone/rclone/fs/pacer.go:88 +0x1c
github.com/rclone/rclone/lib/pacer.(*Pacer).call(0x357ab40, 0x397e3f0, 0x1)
	github.com/rclone/rclone/lib/pacer/pacer.go:197 +0x74
github.com/rclone/rclone/lib/pacer.(*Pacer).CallNoRetry(...)
	github.com/rclone/rclone/lib/pacer/pacer.go:225
github.com/rclone/rclone/backend/azureblob.(*Object).Update(0x3695d10, {0x180d580, 0x36f0f00}, {0x17f501c, 0x351efd0}, {0xa6be9410, 0x3778580}, {0x3605188, 0x1, 0x1})
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1677 +0x758
github.com/rclone/rclone/backend/azureblob.(*Fs).Put(0x3693180, {0x180d580, 0x36f0f00}, {0x17f501c, 0x351efd0}, {0xa6be9410, 0x3778580}, {0x3605188, 0x1, 0x1})
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1110 +0xf8
github.com/rclone/rclone/fs/operations.Copy({0x180d580, 0x36f0f00}, {0x182e48c, 0x3693180}, {0x0, 0x0}, {0x35ec2ca, 0x5}, {0x182ee64, 0x3778580})
	github.com/rclone/rclone/fs/operations/operations.go:472 +0x19ac
github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0x368eb40, {0x180d580, 0x36f0f00}, 0x36f0c30, {0x182e48c, 0x3693180}, 0x0, 0x368ebd8)
	github.com/rclone/rclone/fs/sync/sync.go:402 +0x188
created by github.com/rclone/rclone/fs/sync.(*syncCopyMove).startTransfers
	github.com/rclone/rclone/fs/sync/sync.go:429 +0xcc
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This looks very much like a bug in the azure client library.

Can you reproduce this easily? If so how?

Can you try [the latest beta](https://beta.rclone.org/) also since that has an updated azure client library.

**Comment by robinaly** [comment]
In my opinion, there is no easy way to reproduce as it really seems connected with the azure driver and I am not aware of any possibility to 'mock' the remote part easily. Therefore, anyone wanting to try this would need a azure blob container with write rights. 

I checked the latest beta with the same result (didn't check line by line):

```
2021/11/18 07:34:32 DEBUG : rclone: Version "v1.58.0-beta.5894.51c952654" starting with parameters ["rclone" "sync" "-vv" "test/" "porkvision:/"]
2021/11/18 07:34:32 DEBUG : Creating backend with remote "test/"
2021/11/18 07:34:32 DEBUG : Using config file from "/home/pi/.config/rclone/rclone.conf"
2021/11/18 07:34:32 DEBUG : fs cache: renaming cache item "test/" to be canonical "/home/pi/test"
2021/11/18 07:34:32 DEBUG : Creating backend with remote "porkvision:/"
2021/11/18 07:34:32 DEBUG : fs cache: renaming cache item "porkvision:/" to be canonical "porkvision:"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x577588]

goroutine 51 [running]:
github.com/Azure/azure-storage-blob-go/azblob.blockBlobClient.CommitBlockList({{{{0x0, 0x0}, {0x0, 0x0}, 0x0, {0x0, 0x0}, {0x2cc2670, 0x7}, {0x0, ...}, ...}, ...}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/zz_generated_block_blob.go:82 +0x2f4
github.com/Azure/azure-storage-blob-go/azblob.BlockBlobURL.CommitBlockList({{{{{...}, {...}}}}, {{{{...}, {...}, 0x0, {...}, {...}, {...}, 0x0, {...}, ...}, ...}}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/url_block_blob.go:120 +0x1c0
github.com/Azure/azure-storage-blob-go/azblob.(*copier).close(0x3220d20)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/chunkwriting.go:186 +0x118
github.com/Azure/azure-storage-blob-go/azblob.copyFromReader({0x18fcc0c, 0x32a0a50}, {0x18e485c, 0x3226210}, {0x18f6bb4, 0x2d472c0}, {{0x18fcdec, 0x2c0c690}, 0x0, 0x400000, ...})
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/chunkwriting.go:64 +0x24c
github.com/Azure/azure-storage-blob-go/azblob.UploadStreamToBlockBlob({0x18fcc0c, 0x32a0a50}, {0x18e485c, 0x3226210}, {{{{{...}, {...}}}}, {{{{...}, {...}, 0x0, {...}, ...}, ...}}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/highlevel.go:553 +0xf0
github.com/rclone/rclone/backend/azureblob.(*Object).Update.func1()
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1680 +0x200
github.com/rclone/rclone/fs.pacerInvoker(0x1, 0x1, 0x2c182a0)
	github.com/rclone/rclone/fs/pacer.go:88 +0x1c
github.com/rclone/rclone/lib/pacer.(*Pacer).call(0x2f4c1c0, 0x2c182a0, 0x1)
	github.com/rclone/rclone/lib/pacer/pacer.go:197 +0x74
github.com/rclone/rclone/lib/pacer.(*Pacer).CallNoRetry(...)
	github.com/rclone/rclone/lib/pacer/pacer.go:225
github.com/rclone/rclone/backend/azureblob.(*Object).Update(0x2f20a50, {0x18fcc0c, 0x32a0a50}, {0x18e485c, 0x3226210}, {0xa6a88a58, 0x2f4d1c0}, {0x2d79020, 0x1, 0x1})
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1677 +0x790
github.com/rclone/rclone/backend/azureblob.(*Fs).Put(0x3220b60, {0x18fcc0c, 0x32a0a50}, {0x18e485c, 0x3226210}, {0xa6a88a58, 0x2f4d1c0}, {0x2d79020, 0x1, 0x1})
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1110 +0xf8
github.com/rclone/rclone/fs/operations.Copy({0x18fcc0c, 0x32a0a50}, {0x191eebc, 0x3220b60}, {0x0, 0x0}, {0x2fd8016, 0x5}, {0x191f894, 0x2f4d1c0})
	github.com/rclone/rclone/fs/operations/operations.go:472 +0x19e8
github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0x2e4e780, {0x18fcc0c, 0x32a0a50}, 0x32a09f0, {0x191eebc, 0x3220b60}, 0x32, 0x2e4e818)
	github.com/rclone/rclone/fs/sync/sync.go:402 +0x188
created by github.com/rclone/rclone/fs/sync.(*syncCopyMove).startTransfers
	github.com/rclone/rclone/fs/sync/sync.go:429 +0xcc
```

**Comment by ncw** [comment]
> In my opinion, there is no easy way to reproduce as it really seems connected with the azure driver and I am not aware of any possibility to 'mock' the remote part easily. Therefore, anyone wanting to try this would need a azure blob container with write rights.

That is not a problem - I have access to azure blob storage.

> I checked the latest beta with the same result (didn't check line by line):

So it seemed to crash immediately - is that correct?

> github.com/Azure/azure-storage-blob-go/azblob.blockBlobClient.CommitBlockList({{{{0x0, 0x0}, {0x0, 0x0}, 0x0, {0x0, 0x0}, {0x2cc2670, 0x7}, {0x0, ...}, ...}, ...}}, ...)

There appear to be a load of null pointers here which is a clue

I tried to reproduce this on my raspberry pi 3 but I didn't succeed yet.

Can you try to reproduce it with a single file copy? And if you succeed can you attach the failing file here (or paste a link) please?

So

    rclone copy -vv file porkvision:/

Also I was surprised not to see a log like this - did you edit it out?

    2021/11/18 12:35:17 DEBUG : file: Need to transfer - File not found at Destination

Could you also show your config file please, so run `rclone config show` and redact any secrets by XXX-ing them out - thanks.

**Comment by robinaly** [comment]
Thanks a lot for helping to get to the bottom of this! 

> > In my opinion, there is no easy way to reproduce as it seems connected with the azure driver and I am not aware of any possibility to 'mock' the remote part easily. Therefore, anyone wanting to try this would need an azure blob container with write rights.
> 
> That is not a problem - I have access to azure blob storage.

I guess with my config it becomes clear how to replicate the error. Let me know if anything is missing.


> 
> > I checked the latest beta with the same result (didn't check line by line):
> 
> So it seemed to crash immediately - is that correct?

Yes, timewise it feels as if it was right away.



> Could you also show your config file please, so run `rclone config show` and redact any secrets by XXX-ing them out - thanks.

```
$ rclone config show
```
```
[porkvision]
type = azureblob
sas_url = XXXX
```

> Also I was surprised not to see a log like this - did you edit it out?
> 
> ```
> 2021/11/18 12:35:17 DEBUG : file: Need to transfer - File not found at Destination
> ```
> 

Nope, I posted a straight copy.

> Can you try to reproduce it with a single file copy? And if you succeed can you attach the failing file here (or paste a link) please?
> 
> So
> 
> ```
> rclone copy -vv file porkvision:/
> ```
> 

See output beneath. Still the same error I fear. 

```
2021/11/19 12:21:53 DEBUG : rclone: Version "v1.58.0-beta.5894.51c952654" starting with parameters ["rclone" "copy" "-vv" "file" "porkvision:/"]
2021/11/19 12:21:53 DEBUG : Creating backend with remote "file"
2021/11/19 12:21:53 DEBUG : Using config file from "/home/pi/.config/rclone/rclone.conf"
2021/11/19 12:21:53 DEBUG : fs cache: adding new entry for parent of "file", "/home/pi"
2021/11/19 12:21:53 DEBUG : Creating backend with remote "porkvision:/"
2021/11/19 12:21:53 DEBUG : fs cache: renaming cache item "porkvision:/" to be canonical "porkvision:"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5683b4]

goroutine 1 [running]:
github.com/Azure/azure-storage-blob-go/azblob.blobClient.GetProperties({{{{0x0, 0x0}, {0x0, 0x0}, 0x0, {0x0, 0x0}, {0x38e6c16, 0x6}, {0x0, ...}, ...}, ...}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/zz_generated_blob.go:943 +0x260
github.com/Azure/azure-storage-blob-go/azblob.BlobURL.GetProperties({{{{{...}, {...}, 0x0, {...}, {...}, {...}, 0x0, {...}, {...}, {...}}, ...}}}, ...)
	github.com/Azure/azure-storage-blob-go@v0.14.0/azblob/url_blob.go:187 +0xd8
github.com/rclone/rclone/backend/azureblob.(*Object).readMetaData.func1()
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1457 +0x6c
github.com/rclone/rclone/fs.pacerInvoker(0x1, 0xa, 0x39b6160)
	github.com/rclone/rclone/fs/pacer.go:88 +0x1c
github.com/rclone/rclone/lib/pacer.(*Pacer).call(0x379a480, 0x39b6160, 0xa)
	github.com/rclone/rclone/lib/pacer/pacer.go:197 +0x74
github.com/rclone/rclone/lib/pacer.(*Pacer).Call(0x379a480, 0x39b6160)
	github.com/rclone/rclone/lib/pacer/pacer.go:216 +0x5c
github.com/rclone/rclone/backend/azureblob.(*Object).readMetaData(0x384e9b0)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:1456 +0x21c
github.com/rclone/rclone/backend/azureblob.(*Fs).newObjectWithInfo(0x36ef340, {0xbec666f9, 0x4}, 0x0)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:769 +0xa8
github.com/rclone/rclone/backend/azureblob.(*Fs).NewObject(0x36ef340, {0x18fcc2c, 0x3442090}, {0xbec666f9, 0x4})
	github.com/rclone/rclone/backend/azureblob/azureblob.go:780 +0x34
github.com/rclone/rclone/fs/operations.moveOrCopyFile({0x18fcc2c, 0x3442090}, {0x191eebc, 0x36ef340}, {0x191f858, 0x39ac1c0}, {0xbec666f9, 0x4}, {0xbec666f9, 0x4}, ...)
	github.com/rclone/rclone/fs/operations/operations.go:1819 +0x368
github.com/rclone/rclone/fs/operations.CopyFile(...)
	github.com/rclone/rclone/fs/operations/operations.go:1909
github.com/rclone/rclone/cmd/copy.glob..func1.1()
	github.com/rclone/rclone/cmd/copy/copy.go:88 +0xcc
github.com/rclone/rclone/cmd.Run(0x1, 0x1, 0x23bc4f0, 0x399fea8)
	github.com/rclone/rclone/cmd/cmd.go:255 +0xfc
github.com/rclone/rclone/cmd/copy.glob..func1(0x23bc4f0, {0x381a1b0, 0x2, 0x3})
	github.com/rclone/rclone/cmd/copy/copy.go:84 +0xc8
github.com/spf13/cobra.(*Command).execute(0x23bc4f0, {0x381a198, 0x3, 0x3})
	github.com/spf13/cobra@v1.2.1/command.go:860 +0x644
github.com/spf13/cobra.(*Command).ExecuteC(0x23b9fc8)
	github.com/spf13/cobra@v1.2.1/command.go:974 +0x414
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.2.1/command.go:902
github.com/rclone/rclone/cmd.Main()
	github.com/rclone/rclone/cmd/cmd.go:553 +0x98
main.main()
	github.com/rclone/rclone/rclone.go:14 +0x14
```

**Comment by ncw** [comment]
@robinaly What was in `file`? How big was it?

**Comment by robinaly** [comment]
> @robinaly What was in `file`? How big was it?

Sorry should have specified 

```
touch file
```

So simply an empty file. Maybe not the best test case.

**Comment by ncw** [comment]
I think I've figured out what is happening....

Copying the file to the root will do that using a SAS URL.

You need to include the bucket name, so

    rclone copy -vv file porkvision:bucket

Rclone should warn you about this but it doesn't!

**Comment by robinaly** [comment]
Hi, sorry for the late response. I can confirm that it works with the bucket. Thanks a lot for the support!

**Comment by ncw** [comment]
I've fixed the crash here, so rclone will give a sensible warning rather than crash the Azure SDK

[v1.58.0-beta.5914.c2557cc43.fix-5820-azureblob-sas-crash](https://beta.rclone.org/branch/fix-5820-azureblob-sas-crash/v1.58.0-beta.5914.c2557cc43.fix-5820-azureblob-sas-crash/) on branch [fix-5820-azureblob-sas-crash](https://github.com/rclone/rclone/tree/fix-5820-azureblob-sas-crash) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58
